widgetastic_manageiq.py

```diff
@@ -1773,3 +1774,214 @@ def fill(self, value):
+        objects = [cls(parent=parent, additional_context={'name': name}) for name in names]
+        return objects
+
+    def read(self):
+        return self.quad_icon(self.context['name']).read()
+
+    def fill(self, values):
+        return self.quad_icon(self.context['name']).fill()
+
+    @property
+    def is_displayed(self):
+        try:
+            return super(BaseTileIconItem, self).is_displayed and \
+                self.browser.is_displayed(self.LIST)
+        except NoSuchElementException:
+            return False
+
+
+class ProviderTileIconItem(BaseTileIconItem):
+    pass
+
+
+class BaseListItem(ParametrizedView, ClickableMixin):
+    PARAMETERS = ('name',)
+    TABLE_LOCATOR = ParametrizedLocator('.//table[.//td[normalize-space(.)={name|quote}]]')
+    ROOT = ParametrizedLocator('.//tr[./td[normalize-space(.)={name|quote}]]')
+    parent_table = Table(locator=TABLE_LOCATOR)
+    checkbox = Checkbox(locator='.//input[@type="checkbox"]')
+
+    @property
+    def is_checked(self):
+        return self.checkbox.selected
+
+    def check(self):
+        if not self.is_checked:
+            self.checkbox.click()
+            return True
+        return False
+
+    def uncheck(self):
+        if self.is_checked:
+            self.checkbox.click()
+            return True
+        return False
+
+    @property
+    def href(self):
+        return None
+
+    @property
+    def data(self):
+        row = next(row for row in self.parent_table.rows() if row.name.text == self.context['name'])
+        item_data = {}
+        for col_name in (h for h in self.parent_table.headers if h is not None):
+            item_data[col_name] = row[col_name].text
+        return item_data
+
+    def read(self):
+        return self.is_checked
+
+    def fill(self, values):
+        return self.check(values)
+
+    @classmethod
+    def all_items(cls, parent):
+        br = parent.browser
+        names = [br.get_attribute('title', el) for el in br.elements(cls.ALL)]
+        objects = [cls(parent=parent, additional_context={'name': name}) for name in names]
+        return objects
+
+
+class ProviderListItem(BaseListItem):
+    pass
+
+
+class ProviderItem(View):
+    quad_item = ProviderQuadIconItem
+    list_item = ProviderListItem
+    tile_item = ProviderTileIconItem
+
+    def __init__(self, parent, name, logger=None):
+        View.__init__(self, parent, logger=logger)
+        self.name = name
+
+    def _get_existing_item(self):
+        for item in (self.quad_item, self.tile_item, self.list_item):
+            if item(name=self.name).is_displayed:
+                return item(name=self.name)
+        else:
+            raise NoSuchElementException("Item {name} isn't found on page".format(name=self.name))
+
+    def __getattr__(self, name):
+        if name.startswith('__'):
+            return self.__dict__[name]
+
+        item = self._get_existing_item()
+        if hasattr(item, name):
+            return getattr(item, name)
```

Comment:
The goal here is to return none if name does not exist on the obejct? Wouldn't it be better to raise AttributeError? In that case you could drop the if.